_posts/2017-09-20-mvc-delievery-mechanism-dm.md

```diff
@@ -0,0 +1,178 @@
+
+The main idea behind MVC was to keep the presentation and business logic decoupled from each other.
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-09-20-mvc-delivery-mechanism/mvc.png" alt=“MVC” class="img img-responsive" style="height: 70%; width: 70%;"/>
+</center>
+<br/>
+
+Up until mid 90s, MVC had mainly been used for desktop and embedded applications. [Three-tier architecture][3] (presentation, logic, data) was the reigning architectural style in the 90s, and MVC fitted as a glove. In the late 90s and early 2000s, most companies started migrating their applications to the web and the need to keep the user interface separate from the business logic became even more important. The MVC pattern promotes exactly that and it was a natural step to use MVC for web applications. However, we now had a browser, http calls, and the View would be rendered outside our applications, separately from our Controllers and Model. In order to make things easier for developers, a few frameworks were created. In the Java world, we had MVC [Model 1][4] and [Model 2][5] with [Java Server Pages (JSP)][6] and [Servlets][7]. [Struts][8], the first major Java MVC framework, was created in the early 2000s. From then onwards, we had many attempts to make the communication between the browser (View) and our applications (Controller and Model) seamlessly. Many other frameworks like [Tapestry][21], [Java Server Faces (JSF)][22]and [Spring MVC][23] emerged in the Java world. Other languages and platforms also created their own, [Rails][15] being one of the most successful ones. 
+
+As web applications and frameworks diversified, the MVC pattern also evolved and new variations like [MVA][9], [MVP][10], [MVVM][11] and [PAC][12] emerged. MVC frameworks became popular and as part of their evolution they tried to automate more and more the _manual_ work the developer had to do, including database access. MVC frameworks started automating how [Entities][13] are persisted using [Object-Relational Mapping - ORM][14] techniques or closely integrated with other ORM frameworks. Due to the simplicity of  many web applications, MVC frameworks made it easy to capture data in a web form and store in relational databases. They also made it easy to read data from the database and display on the browser. The price for this automation was a bastardisation of the Model layer, which became synonymous of _Entities_ that represented tables in the database. 
+
+With the Model now associated to Entities and persistency and the View associated to HTML, CSS and JavaScript, developers found no other alternative then to put the business logic of the system in Controller layer. Unfortunately MVC frameworks are one of the main reasons we find Web applications with an anaemic domains (Entities with no behaviour) and fat Controllers - responsible for navigation, managing sessions, parsing JSON and XML, and also with business logic. 
```

Comment:
Maybe throw in a link to Anaemic domain?

You can choose:

- https://en.wikipedia.org/wiki/Anemic_domain_model
- https://www.martinfowler.com/bliki/AnemicDomainModel.html
- https://dzone.com/articles/anaemic-vs-rich-domain-objects-finding-the-balance

Reply:
Good idea. I'll add it